### DISCUSSION TITLE
RUNTIME_DISPATCH_ERROR: Reentrancy detected error

### DISCUSSION THREAD

**Discussion by ariesgun** [aptos-labs/aptos-developer-discussions#414] [Questions] [answered]
### Describe your question in detail.

I have implemented a dispatchable_fungible_asset based on `https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/move-examples/fungible_asset/fa_coin/sources/FACoin.move`.

In my unit test:
```
let fa = primary_fungible_store::withdraw(receiver, metadata, 5);
primary_fungible_store::deposit(to_address, fa);
```

I got this error.
![image](https://github.com/user-attachments/assets/671b7c9a-a2de-43f5-b965-2f9714417277)

According to the documentation, I should be able to use `primary_fungible_store` to deposit or withdraw without changing anything.
What did I miss here?

### Which operating system are you using?

Windows

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by runtian-zhou** [comment]
Thanks for the question! The important thing here is in the dispatch function hook (the withdraw/deposit function you implemented), make sure it will not call into `primary_fungible_store`. The error means that during execution there is a cycle on the call stack (such as  module A calls module B, and module B calls back into module A). Lemme know if you have more questions!

**Comment by ariesgun** [comment]
I see.. Then the first paragraph in [https://aptos.dev/en/build/smart-contracts/fungible-asset#interacting-with-dispatchable-fungible-asset ](here) is not accurate then. Is it possible for users to use `fungible_asset::deposit()` or `fungibe_asset::withdraw()` to transfer the dispatchable fungible asset?

**Reply by kshitijc1506** [reply, marked as answer]
@runtian-zhou perhaps we can make this clearer given the complexity. 

@ariesgun, the issue is possibly that your hooks call into primary_fungible_store.move, please let me know if this is not the case. 

the distinction lies in the fact that:

a) contracts interacting with an asset (say FAcoin) can continue to call into `primary_fungible_store::withdraw/deposit`, since there is no risk of reentrancy. 

i.e. `pfs::withdraw` -> `dfa::withdraw` -> `fa_coin::withdraw` -> `fa::withdraw_with_ref`, so we're good.

b) contracts that define their asset and hook (so you, the person deploying the FACoin module), can't call into _ANY_ function in pfs _WITHIN_ your hook logic. this causes a potential reentrancy, which you're experiencing. 

i.e. `pfs::withdraw` -> `dfa::withdraw` -> `fa_coin::withdraw` -> `pfs::withdraw` (cycle on call stack) (can be any other function in pfs too, that is not inlined). 

basically, its the responsibility of the developer implementing the hooks to prevent cases of cyclical dependencies, not the protocols already interacting with the asset (using primary stores).

**Reply by kshitijc1506** [reply]
also maybe check this discussion out for more context: https://github.com/aptos-labs/aptos-developer-discussions/discussions/296

**Reply by ariesgun** [reply]
Thanks for the detailed explanation. I figured out what caused the reentrancy error.

**Reply by kshitijc1506** [reply]
@ariesgun out of curiosity, what was the issue?

**Reply by ariesgun** [reply]
I called the `primary_fungible_store::deposit/withdraw` in the unit test which is located in the same file as the implementation of dispatchable withdraw/deposit functions. This's what caused the reentrancy error. 
When I called it from a different module, it worked just fine.

**Comment by runtian-zhou** [comment]
No. There is a pretty complex reason for this that. If you are interested in that you should probably look at the [aip](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-73.md). The tl;dr is `fungible_asset::deposit()` only does transfer for non dispatchable assets. You can use `primary_fungible_store::transfer` to transfer both dispatchable and non dispatchable assets.

**Comment by runtian-zhou** [comment]
> Then the first paragraph in [https://aptos.dev/en/build/smart-contracts/fungible-asset#interacting-with-dispatchable-fungible-asset ](https://github.com/aptos-labs/aptos-developer-discussions/discussions/here)is not accurate then

Looking at the doc it does seem accurate?

It does suggest the following:

```
For users using secondary FungibleStore to interact with assets, use dispatchable_fungible_asset::withdraw/deposit instead of fungible_asset::withdraw/deposit to handle assets with registered hooks.
```
